File: 1744-number-of-ways-to-form-a-target-string-given-a-dictionary/1744-number-of-ways-to-form-a-target-string-given-a-dictionary.c
#define MOD 1000000007

int traverse (int level, int index, int freq [] [26], char* target, int** memo, int n, int m);


int numWays (char** words, int wordsSize, char* target) {
    int n = strlen (words [0]);
    int m = strlen (target);
    int freq [n] [26];
    memset(freq, 0, sizeof (freq));

    for (int i = 0; i < wordsSize; i++) {
        for (int j = 0; j < n; j++) {
            freq [j] [words [i] [j] - 'a']++;
        }
    }

    int** memo = (int**) malloc ((n + 1) * sizeof (int*));
    for (int i = 0; i <= n; i++) {
        memo [i] = (int*) malloc ((m + 1) * sizeof (int));
        for (int j = 0; j <= m; j++) {
            memo [i] [j] = -1;
        }
    }

    int result = traverse (0, 0, freq, target, memo, n, m);

    for (int i = 0; i <= n; i++) {
        free (memo [i]);
    }
    free (memo);

    return result;
}

int traverse (int level, int index, int freq [] [26], char* target, int** memo, int n, int m) {
    if (index == m) return 1;
    if (level == n) return 0;
    if (memo [level] [index] != -1) return memo [level] [index];
    
    int ways = traverse (level + 1, index, freq, target, memo, n, m);
    
    if (freq [level] [target [index] - 'a'] > 0) {
        ways = (ways + (long long)freq [level] [target [index] - 'a'] * traverse (level + 1, index + 1, freq, target, memo, n, m)) % MOD;
    }
    
    return memo [level] [index] = ways;
}

File: 1744-number-of-ways-to-form-a-target-string-given-a-dictionary/1744-number-of-ways-to-form-a-target-string-given-a-dictionary.py
class Solution:
    def numWays (self, words : List [str], target : str) -> int:

        MOD = 10 ** 9 + 7
        m, n = len (target), len (words [0])

        freq = [[0] * 26 for _ in range (n)]
        for word in words:
            for i, char in enumerate (word):
                freq [i] [ord (char) - ord ('a')] += 1

        dp = [[0] * (n + 1) for _ in range (m + 1)]
        dp [0] [0] = 1

        for i in range (m + 1):
            for j in range (n):
                dp [i] [j + 1] = (dp [i] [j + 1] + dp [i] [j]) % MOD

                if i < m:
                    char_idx = ord (target [i]) - ord ('a')
                    dp [i + 1] [j + 1] = (dp [i + 1] [j + 1] + dp [i] [j] * freq [j] [char_idx]) % MOD

        return dp [m] [n]